### PR TITLE
Baro: Temperature Related Fixups

### PR DESCRIPTION
**Pull request by WickedShell**
**TL;DR** #5889 is to complicated, this is better, and fixes an over estimation of altitude.

---

This replaces #5889. It was to complicated, and resulted in different behaviors on copter depending upon if you took off with/without GPS (and when you took off).

The core problem with the EAS2TAS calculation that #5889 was attempting to fix was that the source of temperature for EAS2TAS is quite important for ensuring that we aren't providing a reasonable estimate of EAS2TAS. The problem with master is that we double count the affect of any change in altitude, causing the aircraft to change the ratio faster then it actually was.

This PR corrects the use of double counting the temperature for altitude. However while doing that a rethink of how we selected calibration temperatures was needed to further improve accuracies. @priseborough and I discussed a couple of options and the best option was to select an external temperature source if we had one available (airspeed on plane, in future work we can incorporate the concept of external baro sources) if there isn't an available external temperature source we can take the internal one, but given it's tendency to be high we can clip it at a reasonable value (35C covers most tropics/hot environment users safely while still improving the reported altitude for users in cooler regions). The concept of clipping the external temperature source isn't new (we were clipping it at 25C, but we were using it in EAS2TAS which we adjusted to quickly as well)

While implementing the rest of this I came across the fact that the actual relative altitude calculation uses the unclipped barometer temperature as well. This is actually quite poor if your barometer is reading 30 degrees warmer then it actually is (which is very common). This results in a linear error of the aircraft estimating it's higher then it actually is. Using the data collected by the [LOHAN](http://diydrones.com/profiles/blogs/pratchett-soars-to-31-000m-for-pixhawk-test-flight) project provided a great source of data to demonstrate this effect. The red line is the altitude ArduPilot would report using the current temperature selection method, the green line is the reported altitude if we were to clip the temperature sensor at 35C, and the blue line is the true altitude.

figure_1-5
figure_1-6

As you can see there is nearly a 200m altitude error at 2km of altitude. That means we are overestimating our altitude by around 10% and flying lower then intended. On a long terrain following mission this would be especially problematic (Sidenote: GPS altitude is almost always a better altitude source for long missions or terrain following). If we had instead used the 35C clip approach (what this PR presents) we would instead have around 80 meters of altitude error (around 4%).

This PR clips the temperature source used for all altitude estimation at 35C as well. The final step that this PR takes is the GND_TEMP parameter can now be set to provide a true estimate of the ground temperature. During a preflight you can provide the temperature to this parameter (in degrees Celsius) and use your true OAT for the best result. (NOTE: Adjusting the value by a significant amount from the estimate will cause a step input, and upset the EKF. It will recover without any problem, but I don't recommend trying it in flight. For almost all users this is overkill, but for certain use cases the ability.

Because GND_TEMP is no longer the authoritative source on what ground temperature datum is being used this also now logs the current value of the ground temperature data in the BARO message.

Finally I recreated @khancyr's [patch](https://github.com/WickedShell/ardupilot/commit/f7dfac7adebec12b99d44ec8a4de79e20a7c427f) because I couldn't figure out how to easily pull just that patch in from the branch.

### CODE REVIEW

**Comment by khancyr** [comment]
Looks correct ! I think it will be easy to test at 2km, I just need to find those ICBM that are rotting in my garden back ! 
About my patch no problem ! I am not expert in git but using cherry-pick you can take commits from other branch easily (I don't know the cmline but it is easy with gitk ) !

**Comment by rmackay9** [comment]
well, this sounds very promising!
